src/Store/Persistence.re

```diff
@@ -0,0 +1,247 @@
+              | Ok(value) => entry.value = value
+              | Error(error) =>
+                let message = Json.Decode.string_of_error(error);
+                Log.error("Error decoding store file: " ++ message);
+                entry.value = definition.default;
+              }
+            | None => entry.value = definition.default
+            },
+          store.entries,
+        )
+
+      | Error(error) =>
+        let message = Json.Decode.string_of_error(error);
+        Log.error("Error parsing store file: " ++ message);
+      }
+    | exception (Sys_error(message)) =>
+      // Most likely because the file doesn't exist, which is expected, but log it just in case.
+      Log.debug("Unable to read store file: " ++ message)
+    };
+
+    store;
+  };
+
+  let persist = store => {
+    Log.debug("Writing store for " ++ store.name);
+    let entries =
+      store.entries
+      |> List.map((Entry({definition, value})) =>
+           (definition.key, value |> definition.codec.encode)
+         );
+
+    let str = Json.Encode.encode_string(Json.Encode.obj, entries);
+
+    let path =
+      Filesystem.getStoreFolder()
+      |> Result.map(storeFolder => Filename.concat(storeFolder, store.hash))
+      |> ResultEx.flatMap(Filesystem.getOrCreateConfigFolder)
+      |> Result.map(folder => Filename.concat(folder, "store.json"))
+      |> Result.get_ok;
+
+    let outChannel = open_out(path);
+    Printf.fprintf(outChannel, "%s", str);
+    close_out(outChannel);
+  };
+
+  let persistIfDirty = (store, state) => {
+    let isDirty =
+      List.exists(
+        (Entry({definition, value})) =>
+          !definition.codec.equal(value, definition.get(state)),
+        store.entries,
+      );
+
+    if (isDirty) {
+      List.iter(
+        (Entry({definition, _} as entry)) =>
+          entry.value = definition.get(state),
+        store.entries,
+      );
+
+      persist(store);
+    };
+  };
+
+  let get = (definition: Schema.item(_), store) => {
+    let Entry({value, definition: {pipe, _}, _}) =
+      List.find(
+        (Entry({definition: this, _})) => this.key == definition.key,
+        store.entries,
+      );
+
+    Pipe.send(pipe, definition.pipe, value) |> Option.get;
+  };
+};
+
+module Global = {
+  open Oni_Model.State;
+  open Schema;
+
+  let version =
+    define("version", string, BuildInfo.commitId, _ => BuildInfo.commitId);
+  let workspace =
+    define("workspace", option(string), None, state =>
+      Some(state.workspace.workingDirectory)
+    );
+
+  let store =
+    Store.instantiate("global", () =>
+      [Store.entry(version), Store.entry(workspace)]
+    );
+};
+
+module Workspace = {
+  open Schema;
+
+  type state = (Oni_Model.State.t, Revery.Window.t);
+
+  let windowX =
+    define("windowX", int, 0, ((_state, window)) =>
+      Revery.Window.getPosition(window) |> fst
+    );
+  let windowY =
+    define("windowY", int, 0, ((_state, window)) =>
+      Revery.Window.getPosition(window) |> snd
+    );
```

Comment:
Neat. It looks like we can define persistence per-feature, too.

Reply:
I guess with the dependency structure, we may not be able to at this time, as `Persistence` is part of store. Does it make sense to move it to a Feature project, like for Themes or Configuration? (And provide an effect for persistence?)

Reply:
The Persistence infrastructure could be moved to `Core`, but the stores themselves need to have access to `State.t` I think. It shouldn't be a problem to add a `map` function and add feature-defined items to those stores though.

Reply:
I had that in mind, but forgot at some point. Will add to the todo list

Reply:
Done.